File: 2567-closest-nodes-queries-in-a-binary-search-tree/closest-nodes-queries-in-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> closestNodes(TreeNode root, List<Integer> queries) {
        TreeSet<Integer> set = new TreeSet<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()) {
            int size = q.size();
            while(size-- > 0) {
                TreeNode rv = q.poll();
                set.add(rv.val);
                if(rv.left != null) {
                    q.add(rv.left);
                }
                if(rv.right != null) {
                    q.add(rv.right);
                }
            }
        }
        List<List<Integer>> ans = new ArrayList<>();
        for(int query: queries) {
            Integer min = set.floor(query);
            Integer max = set.ceiling(query);
            List<Integer> ll = new ArrayList<>();
            if(min == null) {
                min = -1;
            }
            if(max == null) {
                max = -1;
            }
            ll.add(min);
            ll.add(max);
            ans.add(ll);
        }
        return ans;
    }
}

File: 2567-closest-nodes-queries-in-a-binary-search-tree/closest-nodes-queries-in-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> closestNodes(TreeNode* root, vector<int>& queries) {
        vector<int> v;
        inorder(root, v);
        int q = queries.size();
        vector<vector<int>> ans;
        for(int i = 0; i < q; i++) {
            int lb = lowerBound(v, queries[i]);
            int ub = upperBound(v, queries[i]);
            ans.push_back({lb, ub});
        }
        return ans;
    }
    void inorder(TreeNode* root, vector<int>& v) {
        if(root == nullptr) {
            return;
        }
        inorder(root->left, v);
        v.push_back(root->val);
        inorder(root->right, v);
    }
    int lowerBound(vector<int>& v, int val) {
        int idx = -1;
        int low = 0;
        int high = v.size()-1;
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(v[mid] <= val) {
                idx = v[mid];
                low = mid + 1;
            }
            else {
                high = mid - 1;
            }
        }
        return idx;
    }
    int upperBound(vector<int>& v, int val) {
        int idx = -1;
        int low = 0;
        int high = v.size()-1;
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(v[mid] >= val) {
                idx = v[mid];
                high = mid - 1;
            }
            else {
                low = mid + 1;
            }
        }
        return idx;
    }
};